File: cpp/leetcode/244.cpp
class WordDistance {
public:
    WordDistance(vector<string>& words) {
        for(int i = 0; i < words.size(); i++)
            dict[words[i]].push_back(i);
    }
    int shortest(string word1, string word2) {
        int  i = 0, j = 0, dist = INT_MAX;
        while(i < dict[word1].size() && j <dict[word2].size()) { 
            dist = min(dist, abs(dict[word1][i] - dict[word2][j]));
            dict[word1][i] < dict[word2][j] ? i++: j++;
        }
        return dist;
    }
private:
    unordered_map<string, vector<int>> dict;
};

File: python/leetcode/244.py
class WordDistance(object):
    def __init__(self, words):
        self.d = collections.defaultdict(list)
        for i, word in enumerate(words):
            self.d[word].append(i)

    def shortest(self, w1, w2):
        d1, d2 = self.d[w1], self.d[w2]
        m, n, i, j, res = len(d1), len(d2), 0, 0, 2**31
        while i < m and j < n:
            res = min(res, abs(d1[i] - d2[j]))
            if d1[i] < d2[j]: i += 1
            else: j += 1
        return res